### ISSUE TITLE
Do you need a CRC32 in protocol v10?

### ISSUE THREAD

**Issue by PaulBernier** [WhoSoup/factom-p2p#6]
Can you share a bit more why you feel a CRC32 is needed in your protocol v10? Aren't the error checks of the lower layers good enough for factomd use case?

https://softwareengineering.stackexchange.com/questions/378724/has-crc-check-on-top-of-tcp-any-meaning

**Comment by WhoSoup** [comment]
I went by this: https://stackoverflow.com/questions/3830206/can-a-tcp-checksum-fail-to-detect-an-error-if-yes-how-is-this-dealt-with/15649930#15649930

In a comment below, there's also this quote from a paper
> After an analysis we conclude that the checksum will fail to detect errors for roughly 1 in 16 million to 10 billion packets.

Factom nodes send several million messages a day (which turn into multiple tcp packets), so I figured that throwing CRC32 on top of it wouldn't hurt.

The potential for damage is minimal, since all application messages have their own checks via signatures and hashes, so it affects mostly the p2p messages itself. Things like sending malformed peer share ips, missing handshakes, and other temporary annoyances.

There's an argument for removing it but it's a low-cost check with some upsides.

**Comment by PaulBernier** [comment]
Ok, I don't have a strong argument about remove it. I still quickly checked geth and it doesn't seem they have added such CRC in their p2p messages for instance: https://github.com/ethereum/go-ethereum/blob/master/p2p/message.go#L40

**Comment by WhoSoup** [comment]
That's a good point. I'll drop CRC. For what it's worth, Bitcoin uses the first 4 bytes of SHA256(SHA256(payload)) as checksum.

**Comment by WhoSoup** [comment]
I keep flip-flopping on removing checksums or improving them. The general consensus seems to be that TCP errors do happen but it's hard to know what kind of impact they'll have. For critical messages, additional checks should be used. (http://noahdavids.org/self_published/CRC_and_checksum.html)

The appeal of CRC32 is that it's fast:
```
BenchmarkAdler32-8     	 1414144	       854 ns/op	       8 B/op	       1 allocs/op
BenchmarkCRC32-8       	 7998510	       152 ns/op	       8 B/op	       1 allocs/op
BenchmarkFnv128-8      	  229381	      5199 ns/op	      16 B/op	       1 allocs/op
BenchmarkMD5-8         	  369140	      3140 ns/op	      16 B/op	       1 allocs/op
BenchmarkSHA1-8        	  412408	      2645 ns/op	      32 B/op	       1 allocs/op
BenchmarkSHA256-8      	  205897	      5854 ns/op	      32 B/op	       1 allocs/op
BenchmarkSHA512-8      	  272835	      4391 ns/op	      64 B/op	       1 allocs/op
BenchmarkSHA3256-8     	  154688	      7670 ns/op	     512 B/op	       3 allocs/op
BenchmarkSHA3512-8     	   75841	     16354 ns/op	     576 B/op	       3 allocs/op
BenchmarkWhirlpool-8   	   18272	     65319 ns/op	      64 B/op	       1 allocs/op
```
(https://github.com/SimonWaldherr/golang-benchmarks#hash)

Using SHA1 or SHA256 would be an improvement in terms of security at a slightly higher cost, though the cost is likely negligible.

**Comment by PaulBernier** [comment]
It seems from those references that a CRC32 wouldn't be much benefit because it's already included in the Ethernet layer. So if you want to keep a checksum, SHA1 seems good?

Given the small size of our messages it still seems that the likeliness of an error is very very small. HTTP messages don't have integrity checks for instance. I think checksum are usually useful when downloading entire files. We could also try to open an issue on geth and ask why they decided not to include a checksum to their messages.

If you want to err on the safe side I am fine with adding a SHA1, and we can always re-assess later when we implement the v11 (gob-less protocol)

**Comment by WhoSoup** [comment]
Decided to just remove CRC32 for V10.

Inside the p2p package, we have ping/pong (contents are irrelevant), peer requests, and handshakes. A corrupt message is either going to cause the peer to disconnect (which is "acceptable" behavior) or have malformed data. For ping/pong, malformed data is irrelevant. For handshakes it would be unfortunate but have no dramatic side effects.

Worst case inside p2p would be a peer share having an invalid address. In that case, it would use up one Dial attempt until it times out. Since p2p2 only shares peers that it has an active connection with, this behavior would be the same as trying to connect to a node behind NAT. 

tldr: inside p2p2, the effects of a malformed parcel are very limited.

For factomd, messages are usually either protected by hashes or signed via ed25519. This is more than adequate error protection. A flipped bit would also count as a different message for the duplicate filter. For broadcast messages, all ~16 messages would have to be corrupt to lose a single message.

The list of affected non-broadcast messages are: (not counting data/dblock/missing *responses*) 
https://github.com/FactomProject/factomd/blob/cf81f20045e543772908abbcbc9731214f6c5040/common/constants/constants.go#L84-L85

If one of these messages is missing or broken, factomd is capable of handling it.
